Readme.md
Pomological Colors
================
Garrick Aden-Buie

## Installation

This package isn’t on CRAN, so you’ll need to use the devtools package
to install it.

``` r
# If you don't have devtools installed
install.packages("devtools")

devtools::install_github("gadenbuie/ggpomological")

# To include the vignette
devtools::install_github("gadenbuie/ggpomological", build_vignettes=TRUE)
```

## Introduction

<!-- Links -->

This package provides a ggplot2 theme inspired by the [USDA Pomological
Watercolors collection](https://usdawatercolors.nal.usda.gov/pom) and by
Aron Atkins’s ([@aronatkins](https://twitter.com/aronatkins)) [talk on
parameterized RMarkdown](https://youtu.be/Ol1FjFR2IMU?t=5h21m15s) at
[rstudio::conf 2018](https://www.rstudio.com/conference/).

![](Readme_files/figure-gfm/ggpomological-1.png)<!-- -->

![](Readme_files/pom-examples.jpg)<!-- -->[^1]

## Color Palette

The colors for this theme were drawn from many images from the [USDA
Pomological Watercolors
collection](https://usdawatercolors.nal.usda.gov/pom), I chose just a
few that I thought worked well together for color and fill scales

``` r
scales::show_col(ggpomological:::pomological_palette)
```

![](Readme_files/figure-gfm/unnamed-chunk-1-2.png)<!-- -->

and a few colors for the plot background and decoration

``` r
scales::show_col(unlist(ggpomological:::pomological_base))
```

![](Readme_files/figure-gfm/unnamed-chunk-2-1.png)<!-- -->

I’ve also included a [css file](inst/pomological.css) with the complete
collection of color samples.

## Setup theme and scales

There are three theme-generating functions:

  - `theme_pomological()` sets the plot theme to be representative of
    the paper and styling of the watercolors and includes a
    paper-colored background,

  - `theme_pomological_plain()` has the same styling, just with a
    transparent (or white) background,

  - `theme_pomological_fancy()` has the paper-colored background and
    defaults to a fancy handwritten font ([Homemade
    Apple](https://fonts.google.com/specimen/Homemade+Apple/)).

For color and fill scales, **ggpomological** provides
`scale_color_pomological()` and `scale_fill_pomological()`.

In the future, I might revisit this package to

1.  Increase colors in discrete scale

2.  Setup paired color scales. Lots of great color pairs in the
    extracted colors.

3.  Set up continuous scale colors (we’ll see…)

## Fonts

A handwriting font is needed for the fully authentic pomological look,
and I found a few from Google Fonts that fit the bill.

  - [Mr. De Haviland](https://fonts.google.com/specimen/Mr+De+Haviland)
  - [Homemade Apple](https://fonts.google.com/specimen/Homemade+Apple/)
  - [Marck Script](https://fonts.google.com/specimen/Marck+Script/)
  - [Mr. Bedfort](https://fonts.google.com/specimen/Mr+Bedfort/)

Alternatively, use something like
[calligrapher.com](https://www.calligraphr.com/) to create your own
handwriting font\!

But fonts can be painful in R, so the base functions –
`theme_pomological()` and `theme_pomological_plain()` – don’t change the
font by default. To opt into the full pomological effect, use
`theme_pomological_fancy()` which is just an alias for
`theme_pomological(base_family = "Homemade Apple", base_size = 16)`.

## Add paper background\!

**ggpomological** also provides a function named `paint_pomological`
that uses the
[`magick`](https://cran.r-project.org/web/packages/magick/index.html)
package to add a pomological watercolor paper background and a subtle
texture overlay.

## Demo\!

We’ll need ggplot2 (loaded with **ggpomological**) and dplyr

``` r
library(ggpomological)
library(dplyr)
```

**Warning**: If you don’t have the [above fonts](#fonts) installed,
you’ll get an error message with a lot of warnings when running the
below examples. Just replace `theme_pomological("Homemade Apple", 16)`
with `theme_pomological()` for the basic theme without the crazy fonts.

### Basic iris plot

``` r
# Base plot
basic_iris_plot <- ggplot(iris) +
  aes(x = Sepal.Length, y = Sepal.Width, color = Species) +
  geom_point(size = 2)

# Just your standard Iris plot
basic_iris_plot 
```

![](Readme_files/figure-gfm/plot-demo-2.png)<!-- -->

``` r

# With pomological colors
basic_iris_plot <- basic_iris_plot + scale_color_pomological()
basic_iris_plot
```

![](Readme_files/figure-gfm/plot-demo-3.png)<!-- -->

``` r

# With pomological theme
basic_iris_plot + theme_pomological()
```

![](Readme_files/figure-gfm/plot-demo-4.png)<!-- -->

``` r

# With transparent background
basic_iris_plot + theme_pomological_plain() 
```

![](Readme_files/figure-gfm/plot-demo-5.png)<!-- -->

``` r

# Or with "fancy" pomological settings
pomological_iris <- basic_iris_plot + theme_pomological_fancy()

# Painted!
paint_pomological(pomological_iris, res = 110)
```

![](Readme_files/figure-gfm/plot-demo-1.png)<!-- -->

### Stacked bar chart

``` r
stacked_bar_plot <- ggplot(diamonds) +
  aes(price, fill = cut) +
  geom_histogram(binwidth = 850) + 
  xlab('Price (USD)') + 
  ylab('Count') + 
  ggtitle("ggpomological") +
  scale_x_continuous(label = scales::dollar_format()) +
  scale_fill_pomological()

stacked_bar_plot + theme_pomological("Homemade Apple", 16)
```

![](Readme_files/figure-gfm/plot-bar-chart-2.png)<!-- -->

``` r

paint_pomological(
  stacked_bar_plot + theme_pomological_fancy(),
  res = 110
)
```

![](Readme_files/figure-gfm/plot-bar-chart-1.png)<!-- -->

### Density Plot

``` r
density_plot <- mtcars %>% 
  mutate(cyl = factor(cyl)) %>% 
  ggplot() +
  aes(mpg, fill = cyl, color = cyl)+
  geom_density(alpha = 0.75) + 
  labs(fill = 'Cylinders', colour = 'Cylinders', x = 'MPG', y = 'Density') +
  scale_color_pomological() +
  scale_fill_pomological()

density_plot + theme_pomological("Homemade Apple", 16)
```

![](Readme_files/figure-gfm/plot-density-2.png)<!-- -->

``` r

paint_pomological(
  density_plot + theme_pomological_fancy(),
  res = 110
)
```

![](Readme_files/figure-gfm/plot-density-1.png)<!-- -->

### Points and lines

Data from the Texas Housing

``` r
big_volume_cities <- txhousing %>% 
  group_by(city) %>% 
  summarize(mean_volume = mean(volume, na.rm = TRUE)) %>% 
  arrange(-mean_volume) %>% 
  top_n(length(ggpomological:::pomological_palette)) %>% 
  pull(city)
#> Selecting by mean_volume

full_bar_stack_plot <- txhousing %>% 
  filter(city %in% big_volume_cities) %>% 
  group_by(city, year) %>% 
  summarize(mean_volume = mean(volume, na.rm = TRUE)) %>% 
  ungroup %>% 
  mutate(city = factor(city, big_volume_cities)) %>% 
  ggplot() +
  aes(year, mean_volume, fill = city, group = city) +
  geom_col(position = 'fill', width = 0.9) +
  labs(x = 'City', y = 'Mean Volume', color = 'City') +
  theme(panel.grid.minor.x = element_blank()) +
  scale_fill_pomological()

full_bar_stack_plot + theme_pomological("Homemade Apple", 16)
```

![](Readme_files/figure-gfm/plot-full-bar-stack-2.png)<!-- -->

``` r

paint_pomological(
  full_bar_stack_plot + theme_pomological_fancy(),
  res = 110
)
```

![](Readme_files/figure-gfm/plot-full-bar-stack-1.png)<!-- -->

### One last plot

Using my own handwriting and the `ggridges` package.

``` r
ridges_pomological <- ggplot(diamonds) + 
  aes(x = carat, y = clarity, color = clarity, fill = clarity) + 
  ggridges::geom_density_ridges(alpha = 0.75) + 
  theme_pomological(
    base_family = 'gWriting',
    base_size = 20,
    base_theme = ggridges::theme_ridges()
    ) + 
  scale_fill_pomological() + 
  scale_color_pomological()

paint_pomological(ridges_pomological, res = 110)
#> Picking joint bandwidth of 0.057
```

![](Readme_files/figure-gfm/plot-ridges-1.png)<!-- -->

1.  U.S. Department of Agriculture Pomological Watercolor Collection.
    Rare and Special Collections, National Agricultural Library,
    Beltsville, MD 20705


NEWS.md
# ggpomological 0.1

## v0.1.2

* Copied README to `vignettes/ggpomological`.

## v0.1.1

* Do not mess with fonts as default values of `theme_pomological()`. Created `theme_pomological_fancy()` as wrapper with fancy font defaults instead.


R/extra.R
# # load all colors
# x <- readLines("pomological.css")
# x <- stringr::str_extract(x, "#[0-9a-f]{6}")
# x <- x[!is.na(x)]
# 
# gg_color_hue <- function(n) {
#   hues = seq(15, 375, length = n + 1)
#   hcl(h = hues, l = 65, c = 100)[1:n]
# }
# 
# col2hsv <- function(x) rgb2hsv(col2rgb(x))
# 
# 
# dist2ref_color <- function(color, ref_color) {
#   stopifnot(length(ref_color) == 1)
#   x <- col2hsv(c(color, ref_color)) %>% 
#     t %>%
#     dist %>%
#     as.matrix %>%
#     {tibble('ref_color' = .[length(color) + 1, 1:length(color)])}
#   names(x) <- ref_color
#   x
# }
# 
# compare_to_ggplot <- function(compare_to_ggplot) {
#   pomo_gg <- map_dfr(set_names(compare_to_ggplot), ~ as_tibble(t(col2hsv(.))), .id = "color") %>% 
#     bind_cols(
#       map_dfc(gg_color_hue(set_names(length(compare_to_ggplot))), ~ dist2ref_color(compare_to_ggplot, .))
#     ) %>% 
#     tidyr::gather('ggplot_color', 'dist', -color:-v) %>% 
#     # group_by(color) %>% 
#     # do(dist = min(.$dist), ggplot_color = filter(., dist == min(.$dist))$ggplot_color) %>% 
#     # mutate(dist = map_dbl(dist, ~ .), ggplot_color = map_chr(ggplot_color, ~ .)) %>% 
#     # ungroup %>% 
#     mutate(ggplot_color = factor(ggplot_color, gg_color_hue(length(compare_to_ggplot)))) %>% 
#     arrange(ggplot_color)
#   warning(glue::glue("Palette has {length(compare_to_ggplot)} colors"), call. = FALSE)
#   ggplot(pomo_gg) +
#     aes(x = ggplot_color, y = dist, fill = color, label = color) +
#     geom_label(color = 'white')+
#     # geom_point(shape = 15, size = 5) +
#     scale_fill_identity()
# }
# 
# data_frame(
#   'color' = color_options,
#   # 'group' = sample(c('pomo', 'logical'), length(color_options), replace = TRUE),
#   'x' = pmap_chr(tidyr::crossing(letters, letters), ~paste0(..1, ..2))[1:length(color_options)],
#   'y' = 1:length(color_options)
# ) %>% 
#   ggplot() +
#   aes(x, y, fill = color) +
#   # geom_point(size = 8)+
#   geom_col()+
#   geom_text(aes(label = color), hjust = -0.1) +
#   scale_fill_identity() +
#   coord_flip() +
#   theme_minimal() +
#   #theme_xkcd() +
#   theme(
#     text = element_text(family = 'gWriting', size = 16),
#     plot.background = element_rect(fill = base_colors["paper_light"], color = NA),
#     panel.grid = element_line(color = "#efe1c6"),
#     axis.text = element_text(color = "#655843", size = 14)
#   )
# 
# ordered_plot <- function(color_options, dichromat = FALSE) {
#   if (dichromat) {
#     dichr_type <- sample(c("deutan", "protan", "tritan"), 1)
#     message(glue::glue("color blindness: {dichr_type}"))
#     color_options <- dichromat::dichromat(color_options, dichr_type)
#   }
#   data_frame(
#     color = color_options,
#     x = 1,
#     y = 1:length(color_options)
#   ) %>% 
#     ggplot() +
#     aes(x, y, fill = color, label = color) +
#     geom_tile() +
#     geom_label(color = 'white') +
#     scale_fill_identity()+
#     scale_y_continuous(breaks = 1:length(color_options), labels = 1:length(color_options))+
#     theme_minimal()
# }

R/paint_pomological.R
#' Paint a ggpomological watercolor
#' 
#' Uses [magick] to paint a pomological watercolor. (Paints your plot onto a 
#' pomological watercolor style paper, with texture overlay.)
#' 
#' @references https://usdawatercolors.nal.usda.gov/pom
#' @seealso [theme_pomological]
#' @param pomo_gg A pomologically styled ggplot2 object. See [theme_pomological()]
#' @param width Width of output image in pixels
#' @param height Height of output image in pixels
#' @param pointsize Text size for plot text
#' @param outfile Optional name for output file if you'd like to save the image
#' @param pomological_background Paper image, defaults to paper texture provided
#'   by ggpomological.
#' @param pomological_overlay Overlay texture. Set to `NULL` for no texture.
#' @inheritDotParams magick::image_graph res clip antialias
#' @export
paint_pomological <- function(
  pomo_gg, 
  width = 800, 
  height = 500, 
  pointsize = 16, 
  outfile = NULL,
  pomological_background = pomological_images("background"), 
  pomological_overlay = pomological_images("overlay"),
  ...
) {
  if (!requireNamespace("magick", quietly = TRUE)) {
    stop("The package magick is required for `paint_pomological()`. ",
         "Please install with `install.packages('magick')`")
  }
  if (!file.exists(pomological_background)) {
    warning(paste0("Cannot find file \"", pomological_background, "\""), call. = FALSE)
  }
  
  # Paint figure
  pomo_gg <- pomo_gg + ggplot2::theme(plot.background = ggplot2::element_rect(fill = 'transparent', colour = NA))
  gg_fig <- magick::image_graph(width, height, bg = "transparent", pointsize = pointsize, ...)
  print(pomo_gg)
  dev.off()
  
  if (!is.null(pomological_overlay) && file.exists(pomological_overlay)) {
    pomo_over <- magick::image_read(pomological_overlay)
    pomo_over <- magick::image_resize(pomo_over, paste0(width, "x", height, "!"))
    gg_fig <- magick::image_composite(gg_fig, pomo_over, "blend", compose_args = "15")
  }
  
  # Paint background
  if (file.exists(pomological_background)) {
    pomo_bg <- magick::image_read(pomological_background)
    pomo_bg <- magick::image_resize(pomo_bg, paste0(width, "x", height, "!"))
    pomo_bg <- magick::image_crop(pomo_bg, paste0(width, "x", height))
    
    # Paint figure onto background
    pomo_img <- magick::image_composite(pomo_bg, gg_fig)
  } else pomo_img <- gg_fig
  
  if (!is.null(outfile)) {
    # Do you want your picture framed?
    magick::image_write(pomo_img, outfile)
  }
  pomo_img
}

pomological_images <- function(which = c("background", "overlay")) {
  which <- match.arg(which)
  exts <- c("background" = ".png", "overlay" = ".jpg")
  system.file("images", paste0("pomological_", which, exts[which]),
              package = "ggpomological")
}

R/scale_pomological.R
pomological_palette <- c(
   "#c03728" #red
  ,"#919c4c" #green darkish
  ,"#fd8f24" #orange brighter
  ,"#f5c04a" #yelloww
  ,"#e68c7c" #pink
  ,"#828585" #light grey
  ,"#c3c377" #green light
  ,"#4f5157" #darker blue/grey
  ,"#6f5438" #lighter brown
)

pomological_base <- list(
  "paper"       = "#fffeea",
  "paper_alt"   = "#f8eed1",
  "light_line"  = "#efe1c6",
  "medium_line" = "#a89985",
  "darker_line" = "#6b452b",
  "black"       = "#3a3e3f",
  "dark_blue"   = "#2b323f"
)

#' Pomological Color and Fill Scales
#' 
#' Color scales based on the USDA Pomological Watercolors paintings.
#' 
#' @references https://usdawatercolors.nal.usda.gov/pom
#' @seealso [ggplot2::scale_colour_discrete] [ggplot2::scale_fill_discrete]
#' @inheritDotParams ggplot2::discrete_scale
#' @name scale_pomological
NULL

pomological_pal <- function() scales::manual_pal(pomological_palette)

#' @rdname scale_pomological
#' @export
scale_colour_pomological <- function(...) ggplot2::discrete_scale("colour", "pomological", pomological_pal(), ...)

#' @rdname scale_pomological
#' @export
scale_color_pomological <- scale_colour_pomological

#' @rdname scale_pomological
#' @export
scale_fill_pomological <- function(...) ggplot2::discrete_scale('fill', 'pomological', pomological_pal(), ...)

R/theme_pomological.R
#' Pomological Theme
#' 
#' [ggplot2] plot theme based on the USDA Pomological Watercolors paintings.
#' 
#' @references https://usdawatercolors.nal.usda.gov/pom
#' @seealso [ggplot2::theme]
#' @param base_family Base text family. See **Fonts** in [theme_pomological()] 
#'   for some examples from Google Fonts options, including `"Mr De Haviland"`, 
#'   `"Homemade Apple"`, `"Marck Script"`, and `"Mr. Bedfort"`. For the 
#'   authentic pomological look, use `"Homemade Apple"` or `"Mr De Haviland"`. 
#'   Set to `NULL` or use [theme_pomological_plain()] for no change to fonts.
#' @param base_size Base text size
#' @param text.color Color of all text (except axis text, see `axis.text.color`)
#' @param plot.background.color Color of plot background, passed to `plot.background`
#' @param panel.border.color Color of plot panel border
#' @param with.panel.grid If `FALSE` gridlines in plot are removed
#' @param panel.grid.color,panel.grid.linetype Color and linetype of panel grid, passed to `panel.grid`
#' @param axis.text.color,axis.text.size Color and size of axis text
#' @param base_theme Starting theme of plot, default is 
#'   [ggplot2::theme_minimal()]. Any elements set by `theme_pomological()` will
#'   overwrite the `base_theme` unless the specific parameter is explicitly set
#'   to `NULL`.
#'
#' @section Fonts:
#' Complete the pomological watercolor theme with a handwriting or cursive font.
#' The following fonts from [Google Fonts](https://fonts.google.com) work well.
#' Visit the links below to install on your system.
#' 
#' - [Homemade Apple](https://fonts.google.com/specimen/Homemade+Apple/)
#' - [Mr. De Haviland](https://fonts.google.com/specimen/Mr+De+Haviland)
#' - [Marck Script](https://fonts.google.com/specimen/Marck+Script/)
#' - [Mr. Bedfort](https://fonts.google.com/specimen/Mr+Bedfort/)
#' 
#' Fonts with R are notoriously tricky, so these may not work well for you. If
#' you have installed the fonts but they aren't showing up or working, you can
#' always try running `extrafont::font_import()` or `extrafont::load_fonts()` in
#' the session or RMarkdown document. Or you can use [theme_pomological_plain()].
#'
#' @examples
#' library(ggplot2)
#' basic_iris_plot <- ggplot(iris) +
#'   aes(x = Sepal.Length, y = Sepal.Width, color = Species) +
#'   geom_point(size = 2) +
#'   # with pomological color scale
#'   scale_color_pomological()
#' 
#' # Pomological Theme  
#' basic_iris_plot + 
#'   theme_pomological() 
#' 
#' # With fonts (manual)
#' basic_iris_plot + 
#'   theme_pomological("Homemade Apple", 16)
#'   
#' # Or with fancy alias (same as previous)
#' basic_iris_plot +
#'   theme_pomological_fancy()
#'   
#' # Plain plot without font or background
#' basic_iris_plot +
#'   theme_pomological_plain()
#' 
#' @export
theme_pomological <- function(
  base_family = NULL, 
  base_size = 11,
  text.color = pomological_base$dark_blue,
  plot.background.color = pomological_base$paper,
  panel.border.color = pomological_base$light_line,
  with.panel.grid = FALSE,
  panel.grid.color = pomological_base$light_line,
  panel.grid.linetype = "dashed",
  axis.text.color = pomological_base$medium_line,
  axis.text.size = base_size * 3/4,
  base_theme = ggplot2::theme_minimal()
) {
  if (!is.null(base_family)) check_font(base_family)
  
  base_theme +
    ggplot2::theme(
      text = ggplot2::element_text(
        family = if (!is.null(base_family)) base_family, 
        size = base_size, 
        colour = text.color
      ),
      plot.background = ggplot2::element_rect(
        fill = plot.background.color, 
        colour = NA
      ),
      panel.grid = ggplot2::element_line(
        colour = panel.grid.color,
        linetype = panel.grid.linetype),
      panel.border = ggplot2::element_rect(
        color = panel.border.color,
        fill = NA,
        linetype = "solid",
        size = 0.75
      ),
      panel.grid.major = if (!with.panel.grid) ggplot2::element_blank(),
      panel.grid.minor = ggplot2::element_blank(),
      axis.text = ggplot2::element_text(
        colour = axis.text.color, 
        size = axis.text.size)
    )
}

#' @describeIn theme_pomological Pomological theme with white (transparent) background
#' @export
theme_pomological_nobg <- function(..., plot.background.color = "transparent") {
  theme_pomological(plot.background.color = plot.background.color, ...)
}

#' @describeIn theme_pomological A "plain" pomological theme with white 
#'   background and normal fonts. 
#' @export
theme_pomological_plain <- function(base_family = "", base_size = 11, plot.background.color = "transparent", ...) {
  theme_pomological(base_family, base_size, plot.background.color = plot.background.color, ...)
}

#' @describeIn theme_pomological A "fancy" pomological theme with fancy fonts
#' @export
theme_pomological_fancy <- function(base_family = "Homemade Apple", base_size = 16, ...) {
  theme_pomological(base_family, base_size, ...)
}

font_urls <- data.frame(
  name = c("Mr De Haviland", "Homemade Apple", "Marck Script", "Mr. Bedfort"),
  url  = c(
    "https://fonts.google.com/specimen/Mr+De+Haviland",
    "https://fonts.google.com/specimen/Homemade+Apple/",
    "https://fonts.google.com/specimen/Marck+Script/",
    "https://fonts.google.com/specimen/Mr+Bedfort/"
  )
)

check_font <- function(font_name) {
  if (!requireNamespace("extrafont", quietly = TRUE)) {
    warning("The font \"", font_name, "\" may or may not be installed on your system.",
            "Please install the package `extrafont` if you'd like me to be able to check for you.",
            call. = FALSE)
  } else {
    if (!font_name %in% extrafont::fonts()) {
      if (font_name %in% font_urls$name) {
        warning("Font '", font_name, "' isn't in the extrafonts font list (but it may still work). ", 
                "If recently installed, you can try running `extrafonts::font_import()`. ",
                "To install, visit: ", font_urls[font_urls$name == font_name, "url"],
                call. = FALSE)
      } else {
        warning("Font '", font_name, "' isn't in the extrafonts font list (but it may still work). ", 
                "If recently installed, you can try running `extrafonts::font_import()`. ",
                call. = FALSE)
      }
    }
  }
}
